packages/opentelemetry-context-async-hooks/src/AsyncLocalStorageContextManager.ts

```diff
@@ -30,11 +30,14 @@ export class AsyncLocalStorageContextManager extends AbstractAsyncHooksContextMa
     return this._asyncLocalStorage.getStore() ?? ROOT_CONTEXT;
   }
 
-  with<T extends (...args: unknown[]) => ReturnType<T>>(
+  with<A extends unknown[], F extends (...args: A) => ReturnType<F>>(
     context: Context,
-    fn: T
-  ): ReturnType<T> {
-    return this._asyncLocalStorage.run(context, fn) as ReturnType<T>;
+    fn: F,
+    thisArg?: ThisParameterType<F>,
+    ...args: A
+  ): ReturnType<F> {
+    const cb = thisArg == null ? fn : fn.bind(thisArg);
+    return this._asyncLocalStorage.run(context, cb as any, ...args);
```

Comment:
shouldn't the cb be a type of `F` ?

Reply:
cb is `F | OmitThisParameter<F>` but `this._asyncLocalStorage.run()`  requires `(...args: any[]) => ReturnType<F>`.

Typescript tells then
```
Argument of type 'F | OmitThisParameter<F>' is not assignable to parameter of type '(...args: any[]) => ReturnType<F>'.
  Type 'F' is not assignable to type '(...args: any[]) => ReturnType<F>'.
    Type '(...args: A) => ReturnType<F>' is not assignable to type '(...args: any[]) => ReturnType<F>'.
      Types of parameters 'args' and 'args' are incompatible.
        Type 'any[]' is not assignable to type 'A'.
          'any[]' is assignable to the constraint of type 'A', but 'A' could be instantiated with a different subtype of constraint 'unknown[]'.
```

I tried a little to avoid casting but I was not able to get typescript happy. Must likely because it assumes that our function gets called with unsupported arguments.
I could replace `any` by ` as (...args: unknown[]) => ReturnType<F>` but in the end it just longer and harder to read.